src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -170,35 +170,33 @@ private void OnEventSourceChanged(object? _1, TaggerEventArgs _2)
 
             private void EnqueueWork(bool initialTags)
             {
-                lock (_cancellationSeries)
-                {
-                    try
-                    {
-                        // cancel the last piece of computation work and enqueue the next.  This doesn't apply for the very
-                        // first tag request we make.  We don't want that to be cancellable as we want that result to be 
-                        // shown as soon as possible.
-                        var cancellationToken = initialTags ? _disposalTokenSource.Token : _cancellationSeries.CreateNext();
-
-                        // Continue after the preceeding task unilaterally.  Note that we pass LazyCancellation so that 
-                        // we still wait for that task to complete even if cancelled before we proceed.  This is necessary
-                        // as that prior task may mutate state (even if cancelled) so we cannot proceed until we know it
-                        // is completely done.
-                        _eventWorkQueue = _eventWorkQueue.ContinueWithAfterDelayFromAsync(
-                            async _ =>
-                            {
-                                await this.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                                await RecomputeTagsForegroundAsync(initialTags, cancellationToken).ConfigureAwait(false);
-                            },
-                            cancellationToken,
-                            (int)_dataSource.EventChangeDelay.ComputeTimeDelay().TotalMilliseconds,
-                            TaskContinuationOptions.None,
-                            TaskScheduler.Default).CompletesAsyncOperation(_asyncListener.BeginAsyncOperation(nameof(EnqueueWork)));
-                    }
-                    catch (ObjectDisposedException)
+                using var stateRef = _tagSourceState.TryAddReference();
+
+                // No point proceeding if we've been disposed.
+                if (stateRef is null)
+                    return;
+
+                var state = stateRef.Target;
+
+                // cancel the last piece of computation work and enqueue the next.  This doesn't apply for the very
+                // first tag request we make.  We don't want that to be cancellable as we want that result to be 
+                // shown as soon as possible.
+                var cancellationToken = state.GetCancellationToken(initialTags);
+
+                // Continue after the preceeding task unilaterally.  Note that we pass LazyCancellation so that 
+                // we still wait for that task to complete even if cancelled before we proceed.  This is necessary
+                // as that prior task may mutate state (even if cancelled) so we cannot proceed until we know it
+                // is completely done.
+                state.EventWorkQueue = state.EventWorkQueue.ContinueWithAfterDelayFromAsync(
```

Comment:
❔ Does this state update need to be protected by a lock?

Reply:
yup.  i realized that too later in the day.  it's protected by lock now :)